### PR TITLE
foreign-key-mapping and identity-field

### PR DESCRIPTION
**Pull request by lanxiang1234**
Foreign Key Mapping pattern #1296

Foreign-Key-Mapping

- Maps an association between objects to a foreign key reference between tables.

- Each object contains the database key from the appropriate database table. If two objects are linked together with an association, this association can be replaced by a foreign key in the database.

Pull request description

- tests and three class

> For detailed contributing instructions see https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

### CODE REVIEW

**Comment by lanxiang1234** [comment]
Foreign Key Mapping pattern #1296

File: foreign-key-mapping/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+layout: pattern
+title: Foreign-Key-Mapping
+folder: foreign-key-mapping
+permalink: /patterns/foreign-key-mapping/
+categories: Object-Relational Structural
+tags:
+- foreign-key
```
Comment (iluwatar) [inline foreign-key-mapping/README.md:8]: The available categories and tags are listed in https://github.com/iluwatar/java-design-patterns/wiki/07.-Categories-and-Tags. If not necessary, please don't make new ones. I suggest we use `Structural` category and `Data access` tag.

File: foreign-key-mapping/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+- A Foreign Key Mapping maps an object reference to a foreign key in the database.
+
+- In java programming, objects can refer to each other directly by object references.
+
+- In a relational database, a FOREIGN KEY is a key used to link two tables together.
+
+- In a relational database, a FOREIGN KEY is a field (or collection of fields) in one table that refers to the PRIMARY KEY in another table.
+
+- The table containing the foreign key is called the child table, and the table containing the candidate key is called the referenced or parent table.
```
Comment (iluwatar) [inline foreign-key-mapping/README.md:24]: I suggest we put the key points under separate header. The available headers are listed in https://github.com/iluwatar/java-design-patterns/wiki/02.-Pattern-template.

File: foreign-key-mapping/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+folder: foreign-key-mapping
+permalink: /patterns/foreign-key-mapping/
+categories: Object-Relational Structural
+tags:
+- foreign-key
+---
+
+## Intent
+Maps an association between objects to a foreign key reference between tables.
+
```
Comment (iluwatar) [inline foreign-key-mapping/README.md:13]: Add explanation section here. See the other patterns how it's implemented, for example https://github.com/iluwatar/java-design-patterns/tree/master/abstract-factory

File: foreign-key-mapping/pom.xml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
Comment (iluwatar) [inline foreign-key-mapping/pom.xml:34]: I don't think these are needed. They are specified in the parent pom.xml

File: foreign-key-mapping/pom.xml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (iluwatar) [inline foreign-key-mapping/pom.xml:28]: Add maven-assembly-plugin to enable executing the pattern directly from the command line

File: foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.foreignkey;
+
+
+
+/**
+ * The type Album.
+ * CS304 Issue link: https://github.com/iluwatar/java-design-patterns/issues/1296
```
Comment (iluwatar) [inline foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java:30]: I don't think the issue link is needed here

File: foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+class Album {
+  private long id;
+  private String title;
+  private Artist artist;
```
Comment (iluwatar) [inline foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java:36]: Use Lombok annotations to reduce boilerplate. See https://github.com/iluwatar/java-design-patterns/wiki/18.-Lombok

File: foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java (hunk 6 above)
Comment (iluwatar) [reply foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java:30]: Check elsewhere too

File: foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/App.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.util.LinkedList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * The type App.
+ * CS304 Issue link: https://github.com/iluwatar/java-design-patterns/issues/1296
+ */
```
Comment (iluwatar) [inline foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/App.java:35]: Add here a brief explanation of how the example implements the pattern

File: foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java (hunk 7 above)
Comment (iluwatar) [reply foreign-key-mapping/src/main/java/com/iluwatar/foreignkey/Album.java:36]: Check other classes too

File: foreign-key-mapping/src/test/java/com/iluwatar/foreignkey/ArtistTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class ArtistTest {
+  /**
+   * Artist test.
+   */
+  @Test
+  public void ClassTest() {
+    Artist temp = new Artist(1, "artist");
+    Assert.assertEquals("artist", temp.getName());
+    String t = "artist" + Integer.toString(1);
```
Comment (iluwatar) [inline foreign-key-mapping/src/test/java/com/iluwatar/foreignkey/ArtistTest.java:41]: You can variable type inference (var keyword) in many places

File: identity-field/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
+---
+layout: pattern
+title: Identity-Field
+folder: identity-field
+permalink: /patterns/identity-fields/
+categories: Object-Relational Structural
```
Comment (iluwatar) [inline identity-field/README.md:6]:
```suggestion
categories: Structural
```

File: identity-field/README.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+tags:
+- Identity-Field
```
Comment (iluwatar) [inline identity-field/README.md:8]:
```suggestion
- Data access
```

File: identity-field/README.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+Saves a database ID field in an object to maintain an identity between an in-memory object and a database row.
+
+Representing the Identity Field in an Object:
+
+The simplest form of Identity Field is a field that matches the type of key in the database. Thus, if you use a simple integral key, an integral field will work very nicely.
+
+There are three basic choices to get the key:
+- Get the database to auto-generate
+- Use a GUID
+- Generate your own
```
Comment (iluwatar) [inline identity-field/README.md:21]: Move these under separate heading and add explanation section

File: identity-field/pom.xml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (iluwatar) [inline identity-field/pom.xml:23]: Add maven-assembly-plugin

File: identity-field/pom.xml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+
+
+    </dependencies>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
Comment (iluwatar) [inline identity-field/pom.xml:32]: Remove this part

File: identity-field/src/main/java/com/iluwatar/identityfield/App.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+package com.iluwatar.identityfield;
+
+import java.util.LinkedList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * The type App.
```
Comment (iluwatar) [inline identity-field/src/main/java/com/iluwatar/identityfield/App.java:32]: Explain here how the sample implements the pattern

File: identity-field/src/main/java/com/iluwatar/identityfield/App.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+ */
+@Slf4j
+public class App {
+  /**
+   * The entry point of application.
+   *
+   * @param args the input arguments
+   */
+  public static void main(String[] args) {
+    List<Person> pearsonList = new LinkedList<>();
```
Comment (iluwatar) [inline identity-field/src/main/java/com/iluwatar/identityfield/App.java:42]: Use var

File: identity-field/src/main/java/com/iluwatar/identityfield/Person.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+package com.iluwatar.identityfield;
+
+/**
+ * The type Pearson.
+ */
+public class Person {
+  private long id;
+  private String name;
```
Comment (iluwatar) [inline identity-field/src/main/java/com/iluwatar/identityfield/Person.java:31]: Use Lombok

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by lanxiang1234** [comment]
ok im going to change these problems

**Pull request closed without merging**